core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java

```diff
@@ -20,90 +23,163 @@
 
     private static final Logger LOGGER = Logger.getLogger(IsDockerWorking.class.getName());
 
-    private final boolean silent;
+    private final List<Strategy> strategies;
 
     public IsDockerWorking() {
         this(false);
     }
 
     public IsDockerWorking(boolean silent) {
-        this.silent = silent;
+        this.strategies = List.of(new TestContainersStrategy(silent), new DockerHostStrategy(),
+                new DockerBinaryStrategy(silent));
     }
 
     @Override
     public boolean getAsBoolean() {
-        //remote docker detection
-        //we don't want to pull in the docker API here
-        //so we just see if the DOCKER_HOST is set and we can connect to it
-        //we can't actually verify it is docker listening on the other end
-        String dockerHost = System.getenv("DOCKER_HOST");
-        if (dockerHost != null && !dockerHost.startsWith("unix:")) {
-            try {
-                URI url = new URI(dockerHost);
-                try (Socket s = new Socket(url.getHost(), url.getPort())) {
-                    return true;
-                } catch (IOException e) {
-                    LOGGER.warnf("Unable to connect to DOCKER_HOST URI %s, make sure docker is running on the specified host",
-                            dockerHost);
-                }
-            } catch (URISyntaxException e) {
-                LOGGER.warnf("Unable to parse DOCKER_HOST URI %s, it will be ignored for working docker detection", dockerHost);
+        for (Strategy strategy : strategies) {
+            Result result = strategy.get();
+            if (result != Result.UNKNOWN) {
+                return result == Result.AVAILABLE;
```

Comment:
Are strategies mutually exclusive? i.e what if one strategies return available and the other return unavailable?  Do we consider docker running or not running?

Reply:
It's up to the strategy to return `UNKNOWN` if it doesn't know. I guess we could go through all strategies until we get `AVAILABLE`, that's certainly an option

Reply:
Done

Reply:
Perfect, thanks